### PR TITLE
 ONNX Export Flatten operator

### PR DESCRIPTION
**Pull request by lara-hdr**

### CODE REVIEW

File: test/onnx/test_operators.py [hunk 1]
```diff
@@ -410,9 +410,8 @@ def test_atan(self):
         self.assertONNX(lambda x: x.atan(), x)
 
     def test_flatten(self):
-        # Flatten is a special case of Reshape when the output is a 2-D tensor.
         x = torch.randn(1, 2, 3, 4, requires_grad=True)
-        self.assertONNX(lambda x: x.view(x.size()[0], x.numel() // x.size()[0]), x)
```
Comment (houseroad) [inline test/onnx/test_operators.py]: could you keep the old test case and rename it to something like `test_view_flatten`, and create a new test case for `flattern`

**Review by houseroad — COMMENTED** [review]
Yes, we should encourage user to use flatten function instead.

**Comment by houseroad** [comment]
Also could you fix the lint error?

File: torch/onnx/symbolic.py [hunk 2]
```diff
@@ -1488,3 +1488,23 @@ def log_sigmoid(g, input):
 @parse_args('v')
 def erf(g, input):
     return g.op('Erf', input)
+
+
+@parse_args('v', 'i', 'i')
+def flatten(g, input, start_dim, end_dim) :
+    input_dims = input.type().sizes()
+
+    if end_dim < 0 :
```
Comment (houseroad) [inline torch/onnx/symbolic.py]: We can emit `Flatten` op in the end_dim=-1 case.

Comment (lara-hdr) [reply torch/onnx/symbolic.py]: it seems to me like not all end_dim=-1 cases can be addressed by the Flatten op.
ex : 
input shape : (1,2,3,4)
output = flatten(input,start_dim=2,end_dim=-1)

expected output shape : (1,2,12)
onnx output shape with flatten: (1,24)  (or other 2D shape, depending on the axis chosen)


But the Flatten op can be used when the output shape is 2D. So given the input shape and start_dim/end_dim, we would have to compute the expected output shape, and use Flatten op if the output shape is 2D.

am I missing something?

Comment (houseroad) [reply torch/onnx/symbolic.py]: yeah, you are right, I meant we can emit `Flatten` when possible.

**Review by houseroad — COMMENTED** [review]
Could you add an example which covers the ONNX Flatten path?

**Review by houseroad — APPROVED** [review]
Looks good, thanks!

File: torch/onnx/symbolic.py [hunk 3]
```diff
@@ -1488,3 +1488,31 @@ def log_sigmoid(g, input):
 @parse_args('v')
 def erf(g, input):
     return g.op('Erf', input)
+
+
+@parse_args('v', 'i', 'i')
+def flatten(g, input, start_dim, end_dim) :
+    input_dims = input.type().sizes()
```
Comment (houseroad) [inline torch/onnx/symbolic.py]: Here we should use `dim` to access the dimension information. And after you get the `dim`, please check whether it's `CompleteTensorType` or not, to call `sizes`.

More details, please check https://github.com/pytorch/pytorch/commit/1235aa4fca55cea7c87febcc2761b84bf3114f1a#diff-9488e7f53a53fa19bef4dbae6a0b093a

**Review by houseroad — CHANGES_REQUESTED** [review]
Didn't do well in the first round interview. We should fix this problem.

**Comment by houseroad** [comment]
Could you also add some tests in test/onnx/test_pytorch_onnx_caffe2.py, e.g., https://github.com/pytorch/pytorch/blob/master/test/onnx/test_pytorch_onnx_caffe2.py#L725? The exported model will be executed on caffe2 in such test cases.

**Review by houseroad — APPROVED** [review]

**Comment by houseroad** [comment]
Thanks a lot for updating the pr. appreciate it.
